sklearn/cluster/_hierarchical_fast.pyx

```diff
@@ -29,11 +29,13 @@ from numpy.math cimport INFINITY
 ###############################################################################
 # Utilities for computing the ward momentum
 
-def compute_ward_dist(cnp.ndarray[DOUBLE, ndim=1, mode='c'] m_1,
-                      cnp.ndarray[DOUBLE, ndim=2, mode='c'] m_2,
-                      cnp.ndarray[INTP, ndim=1, mode='c'] coord_row,
-                      cnp.ndarray[INTP, ndim=1, mode='c'] coord_col,
-                      cnp.ndarray[DOUBLE, ndim=1, mode='c'] res):
+cpdef void compute_ward_dist(
+    cnp.float64_t[::1] m_1,
+    cnp.float64_t[:, ::1] m_2,
+    cnp.npy_intp[::1] coord_row,
+    cnp.npy_intp[::1] coord_col,
+    cnp.float64_t[::1] res
+) nogil:
     cdef INTP size_max = coord_row.shape[0]
     cdef INTP n_features = m_2.shape[1]
     cdef INTP i, j, row, col
```

Comment:
for those `cdef`, we could change the `INTP` and `DOUBLE` by the `cnp.npy_intp` and `cnp.float64_t` as well just for consistency in the function.